### PR TITLE
DataFrame conversion tutorial

### PR DESCRIPTION
**Pull request by vincentarelbundock**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [x] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## What is this?

A new tutorial with ultra-simple examples of DataFrame conversion functions using Narwhals.

## Motivation

I maintain a library which uses Polars internally. I want to allow my users to supply any DataFrame format, but I don't want to re-write my library's code, and not all the methods I use are already implemented in Narwhals.

For me, the most attractive application of Narwhals is to use it to "ingest" any DataFrame, and convert them to Polars for internal use. This is the lowest barrier to entry approach to Narwhals, and my guess is that *many* users would like this.

Given the likely popularity of this workflow, I think it deserves its own page and tutorial on the website.

### CODE REVIEW

File: docs/basics/dataframe_conversion.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+```python
+import narwhals as nw
+import polars as pl
+
+
+def df_to_polars(df):
+    out = nw.from_native(df)
+    out = out.to_arrow()
+    return pl.DataFrame(out)
```
Comment (MarcoGorelli) [inline docs/basics/dataframe_conversion.md:31]: can this just be
```python
def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
    return nw.from_arrow(nw.from_native(df), native_namespace=pl)
```
?

Then it'll use the PyCapsule Interface, and depending on the input can avoid going via PyArrow completely

**Review by MarcoGorelli — COMMENTED** [review]
thanks @vincentarelbundock ! just left a comment

it would also be nice to have the code be runnable in the docs, so that it stays up to date - check other docs pages (e.g. docs/how_it_works.md) for examples of how to do that

File: docs/basics/dataframe_conversion.md (hunk 1 above)
Comment (vincentarelbundock) [reply docs/basics/dataframe_conversion.md:31]: This does not appear to work for DuckDB input, which defeats the purpose of a "universal conversion vignette."

My example first requests I call `.to_native()`. When I do that, I get an error.

```python
import narwhals as nw
from narwhals.typing import IntoDataFrame
import duckdb
import polars as pl

df = pl.DataFrame(
  {
    "A": [1, 2, 3, 4, 5],
    "fruits": ["banana", "banana", "apple", "apple", "banana"],
    "B": [5, 4, 3, 2, 1],
    "cars": ["beetle", "audi", "beetle", "beetle", "beetle"],
  }
)
df_duckdb = duckdb.sql("SELECT * FROM df")

def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
  return nw.from_arrow(nw.from_native(df), native_namespace=pl)

df_to_polars(df_duckdb).to_native()
```


Error: 


```python

In [33]: df_to_polars(df_duckdb).to_native()
---------------------------------------------------------------------------
InvalidInputException                     Traceback (most recent call last)
Cell In[33], line 1
----> 1 df_to_polars(df_duckdb).to_native()

Cell In[32], line 2, in df_to_polars(df)
      1 def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
----> 2   return nw.from_arrow(nw.from_native(df), native_namespace=pl)

File ~/Downloads/narwhals/narwhals/functions.py:473, in from_arrow(native_frame, native_namespace)
    468 implementation = Implementation.from_native_namespace(native_namespace)
    470 if implementation is Implementation.POLARS and parse_version(
    471     native_namespace.__version__
    472 ) >= (1, 3):
--> 473     native_frame = native_namespace.DataFrame(native_frame)
    474 elif implementation in {
    475     Implementation.PANDAS,
    476     Implementation.MODIN,
   (...)
    480     # These don't (yet?) support the PyCapsule Interface for import
    481     # so we go via PyArrow
    482     try:

File ~/.virtualenvs/global/lib/python3.12/site-packages/polars/dataframe/frame.py:430, in DataFrame.__init__(self, data, schema, schema_overrides, strict, orient, infer_schema_length, nan_to_null)
    426 elif hasattr(data, "__arrow_c_stream__"):
    427     # This uses the fact that PySeries.from_arrow_c_stream will create a
    428     # struct-typed Series. Then we unpack that to a DataFrame.
    429     tmp_col_name = ""
--> 430     s = wrap_s(PySeries.from_arrow_c_stream(data))
    431     self._df = s.to_frame(tmp_col_name).unnest(tmp_col_name)._df
    433 else:

File ~/Downloads/narwhals/narwhals/dataframe.py:376, in DataFrame.__arrow_c_stream__(self, requested_schema)
    374 native_frame = self._compliant_frame._native_frame
    375 if hasattr(native_frame, "__arrow_c_stream__"):
--> 376     return native_frame.__arrow_c_stream__(requested_schema=requested_schema)
    377 try:
    378     import pyarrow as pa  # ignore-banned-import

InvalidInputException: Invalid Input Error: There is no query result
```

**Comment by vincentarelbundock** [comment]
> thanks @vincentarelbundock ! just left a comment

Your suggested code doesn't work on my computer. I posted a comment above.


> it would also be nice to have the code be runnable in the docs, so that it stays up to date - check other docs pages (e.g. docs/how_it_works.md) for examples of how to do that

I added this to the markdown chunks, but I wasn't sure if the `session` option was necessary, or what I should insert there.

````
```python exec="1" source="above"
2 + 3
```
````

**Comment by MarcoGorelli** [comment]
thanks for updating - the snippet you posted works for me:
```python
In [1]: import narwhals as nw
   ...: from narwhals.typing import IntoDataFrame
   ...: import duckdb
   ...: import polars as pl
   ...:
   ...: df = pl.DataFrame(
   ...:   {
   ...:     "A": [1, 2, 3, 4, 5],
   ...:     "fruits": ["banana", "banana", "apple", "apple", "banana"],
   ...:     "B": [5, 4, 3, 2, 1],
   ...:     "cars": ["beetle", "audi", "beetle", "beetle", "beetle"],
   ...:   }
   ...: )
   ...: df_duckdb = duckdb.sql("SELECT * FROM df")
   ...:
   ...: def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
   ...:   return nw.from_arrow(nw.from_native(df), native_namespace=pl)
   ...:
   ...: df_to_polars(df_duckdb).to_native()
Out[1]:
shape: (5, 4)
┌─────┬────────┬─────┬────────┐
│ A   ┆ fruits ┆ B   ┆ cars   │
│ --- ┆ ---    ┆ --- ┆ ---    │
│ i64 ┆ str    ┆ i64 ┆ str    │
╞═════╪════════╪═════╪════════╡
│ 1   ┆ banana ┆ 5   ┆ beetle │
│ 2   ┆ banana ┆ 4   ┆ audi   │
│ 3   ┆ apple  ┆ 3   ┆ beetle │
│ 4   ┆ apple  ┆ 2   ┆ beetle │
│ 5   ┆ banana ┆ 1   ┆ beetle │
└─────┴────────┴─────┴────────┘
```

could you show the output of `nw.show_versions()` please?

**Comment by vincentarelbundock** [comment]
Sorry for the imprecise report. The `df_to_polars()` works when I call it the first time. But for some reason, it doesn't work when I call it a *second* time on the same DuckDB data frame. 

In contrast, the function I had originally proposed seems to work for multiple calls.

:shrug:

``` python
import narwhals as nw
from narwhals.typing import IntoDataFrame
import duckdb
import polars as pl

df = pl.DataFrame(
  {
    "A": [1, 2, 3, 4, 5],
    "fruits": ["banana", "banana", "apple", "apple", "banana"],
    "B": [5, 4, 3, 2, 1],
    "cars": ["beetle", "audi", "beetle", "beetle", "beetle"],
  }
)
df_duckdb = duckdb.sql("SELECT * FROM df")

def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
  return nw.from_arrow(nw.from_native(df), native_namespace=pl).to_native()

df_to_polars(df_duckdb)

| A   | fruits   | B   | cars     |
|-----|----------|-----|----------|
| i64 | str      | i64 | str      |
| 1   | "banana" | 5   | "beetle" |
| 2   | "banana" | 4   | "audi"   |
| 3   | "apple"  | 3   | "beetle" |
| 4   | "apple"  | 2   | "beetle" |
| 5   | "banana" | 1   | "beetle" |

```
</div>

``` python
df_to_polars(df_duckdb)
```

    InvalidInputException: Invalid Input Error: There is no query result
    [0;31m---------------------------------------------------------------------------[0m
    [0;31mInvalidInputException[0m                     Traceback (most recent call last)
    Cell [0;32mIn[2], line 1[0m
    [0;32m----> 1[0m [43mdf_to_polars[49m[43m([49m[43mdf_duckdb[49m[43m)[49m

    Cell [0;32mIn[1], line 17[0m, in [0;36mdf_to_polars[0;34m(df)[0m
    [1;32m     16[0m [38;5;28;01mdef[39;00m [38;5;21mdf_to_polars[39m(df: IntoDataFrame) [38;5;241m-[39m[38;5;241m>[39m pl[38;5;241m.[39mDataFrame:
    [0;32m---> 17[0m   [38;5;28;01mreturn[39;00m [43mnw[49m[38;5;241;43m.[39;49m[43mfrom_arrow[49m[43m([49m[43mnw[49m[38;5;241;43m.[39;49m[43mfrom_native[49m[43m([49m[43mdf[49m[43m)[49m[43m,[49m[43m [49m[43mnative_namespace[49m[38;5;241;43m=[39;49m[43mpl[49m[43m)[49m[38;5;241m.[39mto_native()

    File [0;32m~/Downloads/narwhals/narwhals/functions.py:473[0m, in [0;36mfrom_arrow[0;34m(native_frame, native_namespace)[0m
    [1;32m    468[0m implementation [38;5;241m=[39m Implementation[38;5;241m.[39mfrom_native_namespace(native_namespace)
    [1;32m    470[0m [38;5;28;01mif[39;00m implementation [38;5;129;01mis[39;00m Implementation[38;5;241m.[39mPOLARS [38;5;129;01mand[39;00m parse_version(
    [1;32m    471[0m     native_namespace[38;5;241m.[39m__version__
    [1;32m    472[0m ) [38;5;241m>[39m[38;5;241m=[39m ([38;5;241m1[39m, [38;5;241m3[39m):
    [0;32m--> 473[0m     native_frame [38;5;241m=[39m [43mnative_namespace[49m[38;5;241;43m.[39;49m[43mDataFrame[49m[43m([49m[43mnative_frame[49m[43m)[49m
    [1;32m    474[0m [38;5;28;01melif[39;00m implementation [38;5;129;01min[39;00m {
    [1;32m    475[0m     Implementation[38;5;241m.[39mPANDAS,
    [1;32m    476[0m     Implementation[38;5;241m.[39mMODIN,
    [0;32m   (...)[0m
    [1;32m    480[0m     [38;5;66;03m# These don't (yet?) support the PyCapsule Interface for import[39;00m
    [1;32m    481[0m     [38;5;66;03m# so we go via PyArrow[39;00m
    [1;32m    482[0m     [38;5;28;01mtry[39;00m:

    File [0;32m~/.virtualenvs/global/lib/python3.12/site-packages/polars/dataframe/frame.py:430[0m, in [0;36mDataFrame.__init__[0;34m(self, data, schema, schema_overrides, strict, orient, infer_schema_length, nan_to_null)[0m
    [1;32m    426[0m [38;5;28;01melif[39;00m [38;5;28mhasattr[39m(data, [38;5;124m"[39m[38;5;124m__arrow_c_stream__[39m[38;5;124m"[39m):
    [1;32m    427[0m     [38;5;66;03m# This uses the fact that PySeries.from_arrow_c_stream will create a[39;00m
    [1;32m    428[0m     [38;5;66;03m# struct-typed Series. Then we unpack that to a DataFrame.[39;00m
    [1;32m    429[0m     tmp_col_name [38;5;241m=[39m [38;5;124m"[39m[38;5;124m"[39m
    [0;32m--> 430[0m     s [38;5;241m=[39m wrap_s([43mPySeries[49m[38;5;241;43m.[39;49m[43mfrom_arrow_c_stream[49m[43m([49m[43mdata[49m[43m)[49m)
    [1;32m    431[0m     [38;5;28mself[39m[38;5;241m.[39m_df [38;5;241m=[39m s[38;5;241m.[39mto_frame(tmp_col_name)[38;5;241m.[39munnest(tmp_col_name)[38;5;241m.[39m_df
    [1;32m    433[0m [38;5;28;01melse[39;00m:

    File [0;32m~/Downloads/narwhals/narwhals/dataframe.py:376[0m, in [0;36mDataFrame.__arrow_c_stream__[0;34m(self, requested_schema)[0m
    [1;32m    374[0m native_frame [38;5;241m=[39m [38;5;28mself[39m[38;5;241m.[39m_compliant_frame[38;5;241m.[39m_native_frame
    [1;32m    375[0m [38;5;28;01mif[39;00m [38;5;28mhasattr[39m(native_frame, [38;5;124m"[39m[38;5;124m__arrow_c_stream__[39m[38;5;124m"[39m):
    [0;32m--> 376[0m     [38;5;28;01mreturn[39;00m [43mnative_frame[49m[38;5;241;43m.[39;49m[43m__arrow_c_stream__[49m[43m([49m[43mrequested_schema[49m[38;5;241;43m=[39;49m[43mrequested_schema[49m[43m)[49m
    [1;32m    377[0m [38;5;28;01mtry[39;00m:
    [1;32m    378[0m     [38;5;28;01mimport[39;00m [38;5;21;01mpyarrow[39;00m [38;5;28;01mas[39;00m [38;5;21;01mpa[39;00m  [38;5;66;03m# ignore-banned-import[39;00m

    [0;31mInvalidInputException[0m: Invalid Input Error: There is no query result

``` python
nw.show_versions()
```


    System:
        python: 3.12.4 (main, Jun  9 2024, 13:30:56) [Clang 15.0.0 (clang-1500.3.9.4)]
    executable: /Users/vincent/.virtualenvs/global/bin/python3
       machine: macOS-14.4-arm64-arm-64bit

    Python dependencies:
         narwhals: 1.10.0
           pandas: 2.2.2
           polars: 1.7.0
             cudf: 
            modin: 
          pyarrow: 17.0.0
            numpy: 2.0.2

**Comment by MarcoGorelli** [comment]
Right, thanks for confirming! there's indeed no guarantee that `__arrow_c_stream__` can be called repeatedly, and this does need documenting, so thanks for highlighting this 🙏 

tbh this pushes me (back) to preferring the name it `from_pycapsule` rather than `from_arrow`, else people expect to be able to call it multiple times just as if they were say passing a pyarrow dataframe to it 🤔 

Anyway, I think we need to document both ways:
- `nw.from_arrow` if you just need to convert once, and this should be more most efficient way
- going via pyarrow, which may be less efficient but can be called liberally

**Comment by vincentarelbundock** [comment]
Sounds good. I edited the file with a short discussion of these issues and the two alternative `df_to_polars()` functions.

**Review by MarcoGorelli — APPROVED** [review]
thanks @vincentarelbundock ! i just made some minor edits, hope it's ok!

**Comment by vincentarelbundock** [comment]
Edits look great, thanks for agreeing to merge this!

Unfortunately, it looks liek the code doesn't execute properly on the site: https://narwhals-dev.github.io/narwhals/basics/dataframe_conversion/

**Comment by MarcoGorelli** [comment]
thanks for spotting this! i think we can just add duckdb to the docs requirements, we'll need to expand duckdb support anyway, there's a lot of demand for that

**Pull request merged**